### PR TITLE
AgentSSH - sign_ssh_data can take optional flags to indicate hashalg

### PR DESCRIPTION
**Pull request by commonism**

### CODE REVIEW

File: paramiko/agent.py [hunk 1]
```diff
@@ -407,12 +410,22 @@ def __str__(self):
     def get_name(self):
         return self.name
 
-    def sign_ssh_data(self, data):
+    def sign_ssh_data(self, data, flags=0):
+        """
+        Sign a blob of data with this private key, and return a `.Message`
+        representing an SSH signature message.
+
+        Use flags to sign using sha2-256 or sha2-512, default is sha1
```
Comment (ploxiln) [inline paramiko/agent.py]: default is sha1 only for rsa keys, not for ecdsa or ed25519, right?

Comment (commonism) [reply paramiko/agent.py]: Yes. flags are not defined for anything but ssh-rsa.

**Comment by ploxiln** [comment]
How might this be used? This seems like just one needed part, at least for a client with an rsa key for user auth.

To auth with key, one calls `Transport.auth_publickey()` (or `SSHClient` calls that for you), and I think that eventually ends up in `AuthHandler._parse_service_accept()` where `sign_ssh_data()` is called (and the key type name is embedded in the message there too). So that would need to know that for "ssh-rsa" keys, if it is using an AgentKey, it can call `sign_ssh_data()` with flag, and change the name from "ssh-rsa" to the appropriate sha2 variant?

Maybe the easier way to use this in paramiko, is to have the agent create 3 keys when it sees an rsa key, one for each hash algo variant. Each AgentKey reports its variant name, and automatically passes the appropriate flag to the agent when `sign_ssh_data()` is called, so `sign_ssh_data()` can keep the same method signature as on every other key type.

Just thinking out-loud ...

**Comment by commonism** [comment]
I'm not looking for ssh authentication with this one. I just need to sign.
Sign data without having key and data in a single place by forwarding access to the key via ssh.

Used here:
https://github.com/m32/endesive/pull/43

**Comment by ploxiln** [comment]
I see. Anyway, this is a useful reference/hint for implementing SSH auth, thanks.

**Comment by bitprophet** [comment]
See #1643 (though I ended up writing my own branch for this). Releasing soon. Thanks!

**Pull request closed without merging**